File: client/protocol/exceptions.cpp
#include "exceptions.hpp"

#include <string>

namespace messageu {
namespace protocol {
namespace exceptions {
GeneralException::GeneralException(const std::string &reason)
    : std::runtime_error(reason) {}

UnexpectedReponseError::UnexpectedReponseError(const types::Code &expected_code,
                                               const types::Code &received_code)
    : GeneralException("Expected response code " + std::string(expected_code) +
                       ", but received " + std::string(received_code)) {}

PayloadMismatch::PayloadMismatch(const types::PayloadSize &expected_size,
                                 const types::PayloadSize &received_size)
    : GeneralException(
          "Expected payload of size " + std::string(expected_size) +
          ", but received payload of size" + std::string(received_size)) {}

ContentMismatch::ContentMismatch()
    : GeneralException("The content size does not match the payload size") {}

ContentSizeLimit::ContentSizeLimit(std::uintmax_t max_size_bytes,
                                   std::uintmax_t received_size_bytes)
    : GeneralException("Tried to send content of size " +
                       std::to_string(received_size_bytes) +
                       "bytes, but the limit is " +
                       std::to_string(max_size_bytes) + "bytes") {}

GeneralError::GeneralError()
    : GeneralException("Server responded with an error") {}
}  // namespace exceptions
}  // namespace protocol
}  // namespace messageu

File: server/protocol/exceptions.py
"""Internal exceptions for rules the MessageU protocol enforces

"""

from protocol import types


class ProtocolError(RuntimeError):
    """Error interface"""


class MessageTypeError(ProtocolError):
    """Received an unknown message type"""
    def __init__(self, message_type: types.MessageType):
        super().__init__('received a message of unknown type')
        self._message_type = message_type

    def __str__(self):
        return '%s (%s)' % (super(), self._message_type.value)


class MessageSizeMismatch(ProtocolError):
    """The message size does not match the payload size"""
    def __init__(self, expected_size: types.PayloadSize,
                 actual_size: types.PayloadSize):
        super().__init__(
            'expected payload size does not match actual payload size')
        self._expect_size = expected_size
        self._actual_size = actual_size

    def __str__(self):
        return '%s (expected: %s, actual: %s)' % (super(), self._expect_size,
                                                  self._actual_size)
